### ISSUE TITLE
lstm4cryptoCODE

### ISSUE THREAD

**Issue by Dzzz339** [Dzzz339/lstm4cryptoCODE#1]
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from binance.client import Client
import time
from collections import Counter
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import confusion_matrix, classification_report
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import GRU, Dense, Dropout, BatchNormalization
from tensorflow.keras.callbacks import EarlyStopping
import tensorflow as tf
from imblearn.over_sampling import SMOTE

# Функция для получения данных с Binance
def get_binance_data(symbol, interval, lookback_days=30):
    client = Client()  # Убедитесь, что у вас есть API ключи, если они требуются
    all_data = pd.DataFrame()
    end_time = pd.Timestamp.now()
    start_time = end_time - pd.Timedelta(days=lookback_days)

    while start_time < end_time:
        start_str = str(int(start_time.timestamp() * 1000))
        klines = client.get_historical_klines(symbol, interval, start_str, limit=1000)
        if not klines:
            break
        data = pd.DataFrame(klines, columns=[
            "timestamp", "open", "high", "low", "close", "volume",
            "close_time", "quote_asset_volume", "trades",
            "taker_buy_base", "taker_buy_quote", "ignore"
        ])
        data["timestamp"] = pd.to_datetime(data["timestamp"], unit="ms")
        data.set_index("timestamp", inplace=True)
        data = data.apply(pd.to_numeric, errors="coerce")
        all_data = pd.concat([all_data, data])
        start_time = data.index[-1] + pd.Timedelta(minutes=5)
        time.sleep(0.5)
    return all_data.sort_index().ffill()

# Функция для добавления признаков
def add_features(df, btc_df):
    df = df.reindex(btc_df.index, method='ffill')
    btc_df = btc_df.reindex(df.index, method='ffill')
    df["Returns"] = df["close"].pct_change()
    df['SMA_10'] = df['close'].rolling(window=10).mean()
    df['EMA_10'] = df['close'].ewm(span=10, adjust=False).mean()
    delta = df['close'].diff()
    gain = (delta.where(delta > 0, 0)).rolling(window=14).mean()
    loss = (-delta.where(delta < 0, 0)).rolling(window=14).mean()
    rs = gain / loss
    df['RSI_14'] = 100 - (100 / (1 + rs))
    df['Volatility'] = df['close'].rolling(window=10).std()
    df['Price_Change'] = df['close'].diff()
    df['BTC_Correlation'] = df['close'].rolling(window=10).corr(btc_df['close'])
    volume_window = 12
    df["BTC_DOGE_Volume_Ratio"] = (btc_df["volume"].rolling(volume_window).mean() / 
                                   df["volume"].rolling(volume_window).mean())
    df["hour"] = df.index.hour
    df["minute"] = df.index.minute
    df["day_part"] = np.where(df['hour'] < 12, 0, 1)
    df["DOGE_BTC_Ratio"] = df["close"] / btc_df["close"]
    df["Target"] = np.where(df["close"].shift(-1) > df["close"] * 1.005, 1, 0)
    df["spread"] = df["high"] - df["low"]
    df["OTKLSPREAD"] = (df["spread"] - df["spread"].mean()) / df["spread"].std()

    # Добавляем MACD
    exp1 = df['close'].ewm(span=12, adjust=False).mean()
    exp2 = df['close'].ewm(span=26, adjust=False).mean()
    df['MACD'] = exp1 - exp2
    df['MACD_Signal'] = df['MACD'].ewm(span=9, adjust=False).mean()

    # Добавляем Bollinger Bands
    df['BB_Middle'] = df['close'].rolling(window=20).mean()
    df['BB_Std'] = df['close'].rolling(window=20).std()
    df['BB_Upper'] = df['BB_Middle'] + 2 * df['BB_Std']
    df['BB_Lower'] = df['BB_Middle'] - 2 * df['BB_Std']

    df.dropna(inplace=True)
    return df

# Загрузка данных
print("=== Загрузка данных ===")
doge = get_binance_data("DOGEUSDT", Client.KLINE_INTERVAL_5MINUTE)
btc = get_binance_data("BTCUSDT", Client.KLINE_INTERVAL_5MINUTE)

# Добавление признаков
print("=== Обработка данных ===")
feature_doge_5min = add_features(doge, btc)

# Анализ целевой переменной
print("=== Анализ данных ===")
target = feature_doge_5min["Target"]
class_counts = Counter(target)
print("Распределение классов:")
for class_name, count in class_counts.items():
    percentage = count / len(target) * 100
    print(f"Класс {class_name} ({'Рост' if class_name == 1 else 'Падение/Стабильность'}): "
          f"{count} записей ({percentage:.2f}%)")

print("\nПервые 5 строк обработанных данных:")
print(feature_doge_5min[["close", "Target", "SMA_10", "EMA_10", "RSI_14", "MACD", "BB_Upper", "BB_Lower"]].head().to_string())

# Подготовка данных
LOOKBACK = 24
TEST_SIZE = 0.2

train_size = int(len(feature_doge_5min) * (1 - TEST_SIZE))
train_data = feature_doge_5min[:train_size]
test_data = feature_doge_5min[train_size:]

scaler = MinMaxScaler()
scaled_train = scaler.fit_transform(train_data.drop("Target", axis=1))
scaled_test = scaler.transform(test_data.drop("Target", axis=1))

# Применяем SMOTE для оверсэмплинга
smote = SMOTE(random_state=42)
X_train_2d = scaled_train
y_train = train_data["Target"].values
X_train_2d_resampled, y_train_resampled = smote.fit_resample(X_train_2d, y_train)

# Преобразуем обратно в 3D для GRU
def create_sequence(data, targets, lookback):
    X, y = [], []
    print(f"Создание последовательностей: len(data)={len(data)}, lookback={lookback}")
    if len(data) <= lookback:
        print("Ошибка: данных недостаточно для создания последовательностей")
        return np.array([]), np.array([])
    for i in range(lookback, len(data)):
        X.append(data[i - lookback:i])
        y.append(targets[i])
    X, y = np.array(X), np.array(y)
    print(f"Размер X: {X.shape}, Размер y: {y.shape}")
    return X, y

# Создаем последовательности после SMOTE
X_train_val, y_train_val = create_sequence(X_train_2d_resampled, y_train_resampled, LOOKBACK)
X_test, y_test = create_sequence(scaled_test, test_data["Target"].values, LOOKBACK)

if X_train_val.size == 0 or X_test.size == 0:
    print("Ошибка: последовательности не созданы. Уменьшите LOOKBACK или увеличьте объем данных.")
    exit()

# Разделение на тренировочную и валидационную выборки
val_size = int(len(X_train_val) * 0.2)
X_train = X_train_val[:-val_size]
y_train = y_train_val[:-val_size]
X_val = X_train_val[-val_size:]
y_val = y_train_val[-val_size:]
print(f"\nРазмеры наборов данных:")
print(f"Тренировочный: X={X_train.shape}, y={y_train.shape}")
print(f"Валидационный: X={X_val.shape}, y={y_val.shape}")
print(f"Тестовый: X={X_test.shape}, y={y_test.shape}")

# Создание улучшенной модели с 1024 нейронами на входе
def create_optimized_model(input_shape):
    model = Sequential()
    model.add(GRU(512,  return_sequences=True, input_shape=input_shape, 
                  kernel_initializer="glorot_uniform", dropout=0.2, recurrent_dropout=0.2))
    model.add(Dropout(0.2))
    model.add(GRU(256, return_sequences=True, kernel_initializer="glorot_uniform"))
    model.add(Dropout(0.2))
    model.add(GRU(128, return_sequences=True, kernel_initializer="glorot_uniform", dropout=0.2))
    model.add(Dropout(0.2))
    model.add(GRU(96, return_sequences=False))
    model.add(Dropout(0.3))
    model.add(Dense(64, activation="relu"))
    model.add(BatchNormalization())
    model.add(Dropout(0.2))
    model.add(Dense(1, activation="sigmoid"))
    
    optimizer = tf.keras.optimizers.Adam(learning_rate=0.001)
    model.compile(optimizer=optimizer, loss="binary_crossentropy", metrics=[
        "accuracy", 
        tf.keras.metrics.Precision(name="precision"),
        tf.keras.metrics.Recall(name="recall"), 
        tf.keras.metrics.AUC(name="auc")
    ])
    return model

model = create_optimized_model((LOOKBACK, X_train.shape[2]))
print("=== Модель создана ===")

# Вычисляем class_weight для взвешивания классов
class_weight = {0: 1.0, 1: (len(y_train) - sum(y_train)) / sum(y_train)}
print(f"Class weights: {class_weight}")

# Обучение модели
Epochs = 100
Batch_size = 32
early_stopping = EarlyStopping(monitor="val_loss", patience=10, restore_best_weights=True)

history = model.fit(
    X_train, y_train,
    validation_data=(X_val, y_val),
    epochs=Epochs,
    verbose=1,
    batch_size=Batch_size,
    callbacks=[early_stopping],
    class_weight=class_weight
)

# Визуализация истории обучения
plt.figure(figsize=(12, 5))
plt.subplot(1, 2, 1)
plt.plot(history.history['loss'], label='Тренировочная потеря')
plt.plot(history.history['val_loss'], label='Валидационная потеря')
plt.title('Динамика потерь')
plt.xlabel('Эпоха')
plt.ylabel('Потеря')
plt.legend()
plt.grid()

plt.subplot(1, 2, 2)
plt.plot(history.history['accuracy'], label='Тренировочная точность')
plt.plot(history.history['val_accuracy'], label='Валидационная точность')
plt.title('Динамика точности')
plt.xlabel('Эпоха')
plt.ylabel('Точность')
plt.legend()
plt.grid()
plt.tight_layout()
plt.show()

# Оценка модели с порогом 0.005
def evaluate_model(model, X_test, y_test, features, threshold=0.005):
    test_pred = model.predict(X_test)
    y_pred = (test_pred > threshold).astype(int)
    
    print("\n=== Результаты оценки модели ===")
    print(f"Порог классификации: {threshold:.2f}")
    print("\nОтчет о классификации:")
    print(classification_report(y_test, y_pred, target_names=["Падение/Стабильность", "Рост"]))
    
    cm = confusion_matrix(y_test, y_pred)
    plt.figure(figsize=(6, 4))
    sns.heatmap(cm, annot=True, fmt="d", cmap="Blues",
                xticklabels=["Падение/Стаб.", "Рост"],
                yticklabels=["Падение/Стаб.", "Рост"])
    plt.xlabel("Предсказание")
    plt.ylabel("Реальность")
    plt.title(f"Матрица ошибок (Порог = {threshold:.2f})")
    plt.show()
    
    plt.figure(figsize=(10, 5))
    sns.histplot(test_pred, bins=50, kde=True, color="purple")
    plt.axvline(threshold, color="red", linestyle="--", label=f"Порог ({threshold:.2f})")
    plt.title("Распределение предсказанных вероятностей")
    plt.xlabel("Вероятность роста")
    plt.ylabel("Частота")
    plt.legend()
    plt.grid()
    plt.show()

evaluate_model(model, X_test, y_test, features=feature_doge_5min.drop("Target", axis=1), threshold=0.005)
